jax/interpreters/xla.py

```diff
@@ -233,17 +233,38 @@ def device_persistent_result_handler(result_shape):
   else:
     raise TypeError(t)
 
-def pyval_result_handler(result_shape):
+def _pyval_result_prefetcher(result_shape):
+  t = type(result_shape)
+  if t is ResultArray:
+    return lambda buf: _copy_to_host_async(buf)
+  elif t is ResultTuple:
+    _, result_shapes = result_shape
+    handlers = list(map(_pyval_result_prefetcher, result_shapes))
+    def f(buf):
+      for h, b in zip(handlers, buf.destructure()):
+        h(b)
+    return f
+  else:
+    raise TypeError(t)
+
+def _pyval_result_fetcher(result_shape):
   t = type(result_shape)
   if t is ResultArray:
     return lambda buf: buf.to_py()
   elif t is ResultTuple:
     _, result_shapes = result_shape
-    handlers = list(map(pyval_result_handler, result_shapes))
+    handlers = list(map(_pyval_result_fetcher, result_shapes))
     return lambda buf: JaxTuple(h(b) for h, b in zip(handlers, buf.destructure()))
   else:
     raise TypeError(t)
 
+def pyval_result_handler(result_shape):
```

Comment:
Exceedingly minor refactoring idea: maybe we could have `prefetcher` return a reference to the buf (i.e. return its input). Then we could write the latter part of this function as `fetcher(prefetcher(buf))`, and we might even want to inline this code at the two call sites in `result_handler` (though I'm less sure about that second part).

Reply:
Moot given prefetcher is now gone.